File: Lab2/recursiveHanoi.java
/*
Author by Rob Mullins

Below is the recursive implimentation of the TOH problem.
Here the base case is of course when the number of disks gets to 1
we continue to move the rods back and forth recursively
*/
package Lab2;

public class recursiveHanoi{
    private Lab2.fileTasks write = new fileTasks();
    private Long writeTime = (long)0;
    
    public Long solve(int num, char fromRod, char toRod, char spareRod, String fileName){
        //Here we have the base case where you just move it once.
        Lab2.ticToc time = new ticToc();
        
        Long start = time.tic();
        if(num == 1){
            // Adding up the time it takes to write to file to subtract later
            writeTime = write.writeFile(fileName, "Move disk 1 from rod " +  fromRod + " to rod " + toRod + "\n") + writeTime;
        }
        else{
            // Here we will recursively call the solve function to go down
            // from here it will count down the disks
            solve(num-1, fromRod, spareRod,toRod,fileName);
            writeTime = (write.writeFile(fileName, "Move disk " + num + " from rod " +  fromRod + " to rod " + toRod + "\n")) + writeTime;
            solve(num-1, spareRod, toRod, fromRod,fileName); 

        }
        Long end = time.toc();
        Long total = (end - start) - (writeTime);
        return total;

    }

}

File: recursiveHanoi.py
import time
from fileTasks import *
class recursiveHanoi:
    def __init__(self,fileName):
        self.f = fileTask()
        self.fileName = fileName
        self.sumWriteTime = int(0)
        self.sumRunTime = int(0)

    def recHanoi(self, disk,fromRod,toRod,spareRod):
        start = time.perf_counter_ns()
        recursive = recursiveHanoi(self.fileName)
        if disk ==1 :
            self.sumWriteTime = self.f.fileWrite(self.fileName,"Move disk 1 from rod "+fromRod+" to rod "+toRod + "\n") + self.sumWriteTime
        else:

            recursive.recHanoi(disk-1,fromRod,spareRod,toRod)
            self.sumWriteTime = self.f.fileWrite(self.fileName,"Move Disk "+str(disk)+" from rod "+fromRod+" to rod " + toRod+ "\n") + self.sumWriteTime
            recursive.recHanoi(disk-1,spareRod,toRod,fromRod) 

        end = time.perf_counter_ns()
        self.sumRunTime = (end-start) + self.sumRunTime 
        return self.sumRunTime - self.sumWriteTime
